### ISSUE TITLE
long latency node join 5 nodes cluster failure

### ISSUE THREAD

**Issue by rockkoca** [ubuntu/microk8s#2033]
I tried to join a node to my existing cluster from a different reagin which has 100ms latency. The API server of the new node keeps restarting. 

But, if I join another empty one node cluster in the same place with the 5 nodes cluster. It could work.

It seems the API server waited too short time before the deadline is reached? 

```
Feb 19 05:01:15 de1 microk8s.daemon-apiserver[13747]: W0219 05:01:15.220327   13747 authentication.go:519] AnonymousAuth is not allowed with the AlwaysAllow authorizer. Resetting AnonymousAuth to false. You should use a different authorizer
Feb 19 05:02:15 de1 microk8s.daemon-apiserver[13747]: Error: context deadline exceeded
Feb 19 05:02:15 de1 systemd[1]: snap.microk8s.daemon-apiserver.service: Main process exited, code=exited, status=1/FAILURE
Feb 19 05:02:15 de1 systemd[1]: snap.microk8s.daemon-apiserver.service: Failed with result 'exit-code'.
Feb 19 05:02:15 de1 systemd[1]: snap.microk8s.daemon-apiserver.service: Scheduled restart job, restart counter is at 10.
```

**Comment by rockkoca** [comment]
It restarts every 60 seconds.

I checked the doc. The only argument that is 1m0s is '--request-timeout'.
But, it does not help even if I changed it to 15m0s. It still restarts every 60 seconds. Does anybody know that which argument can make the timeout longer?

https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/

**Comment by balchua** [comment]
Just checking here, does joining command completes or no?  When you do `kubectl get no` do you see the node or not?

**Comment by rockkoca** [comment]
> kubectl get no

microk8s.kubectl get no
The connection to the server 127.0.0.1:16443 was refused - did you specify the right host or port?

But, I can see 3 master nodes in /var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml.

**Comment by balchua** [comment]
Confirming just so i understand, you are on 2 regions both with 100ms+ latency.
When you build a 3 node HA cluster even if they are on different regions, they are working.

But once you increased the number of nodes to 5 spread across 2 regions, the cluster stopped working.

**Comment by rockkoca** [comment]
> Confirming just so i understand, you are on 2 regions both with 100ms+ latency.
> When you build a 3 node HA cluster even if they are on different regions, they are working.
> 
> But once you increased the number of nodes to 5 spread across 2 regions, the cluster stopped working.

I tried forming two nodes as a HA cluster in 2 regions with 100ms+ delay. It worked. Both nodes are empty new nodes.

I have 5 nodes in one region already with a lot of pods running. And, after I tried to join another node to this cluster in a different region with 100ms+ delay. It failed. Just keeps restarting the API server every 60 seconds.

**Comment by balchua** [comment]
Ok understood now thanks.  Well 2 nodes doesn't really form an HA cluster.  Although i am not very familiar with dqlite, i am not sure if the raft logs are propagated to node 2.

If you try to form a 3 node empty cluster across the 2 region (i.e. containing only the default contents) does it work?  My guess is, yes.

But as you pointed out, your existing cluster already contains lots of data, maybe the log replication is causing those deadline exceeded errors.

**Comment by rockkoca** [comment]
> Ok understood now thanks. Well 2 nodes doesn't really form an HA cluster. Although i am not very familiar with dqlite, i am not sure if the raft logs are propagated to node 2.
> 
> If you try to form a 3 node empty cluster across the 2 region (i.e. containing only the default contents) does it work? My guess is, yes.
> 
> But as you pointed out, your existing cluster already contains lots of data, maybe the log replication is causing those deadline exceeded errors.

Thanks,

Three new nodes in these two regions also work.

**Comment by balchua** [comment]
Hi @rockkoca i am guessing you want to have your HA control plane across 2 regions.

But if you only want to host your workload on a different region than your control plane, i wrote some instructions on how to do so.

https://discuss.kubernetes.io/t/adding-a-worker-node-only-with-microk8s/14801

Its a bit involved though but definitely scriptable. 😊

**Comment by rockkoca** [comment]
> Hi @rockkoca i am guessing you want to have your HA control plane across 2 regions.
> 
> But if you only want to host your workload on a different region than your control plane, i wrote some instructions on how to do so.
> 
> https://discuss.kubernetes.io/t/adding-a-worker-node-only-with-microk8s/14801
> 
> Its a bit involved though but definitely scriptable. 😊

Thanks, This seems to be a great idea. I will take a look

**Comment by balchua** [comment]
Let me know if you manage to set it up or anything that can help improve the doc. 
Thanks.

**Comment by rockkoca** [comment]
> Let me know if you manage to set it up or anything that can help improve the doc.
> Thanks.

Thanks for asking.

It's pretty complex. I haven't had time to do it. I will try if I have time later.

If we could specify worker node only when joining a cluster in the future,  it will be great.

**Comment by balchua** [comment]
Thanks @rockkoca . Take your time.  😊

**Comment by balchua** [comment]
This maybe related to this issue. 
https://github.com/canonical/dqlite/issues/282

**Comment by rockkoca** [comment]
> This maybe related to this issue.
> [canonical/dqlite#282](https://github.com/canonical/dqlite/issues/282)

Yes. This is not about latency. The long latency node finally joined after so many days without any action done by me.

However, one of my nodes stopped working. It first started with 'Temporary failure in name resolution errors'. After I reinstall microk8s and tried to rejoin it, the same problem happened. 

`Mar 05 21:23:59 k-nas2 microk8s.daemon-apiserver[249953]: Error: context deadline exceeded`

**Comment by rockkoca** [comment]
> This maybe related to this issue.
> [canonical/dqlite#282](https://github.com/canonical/dqlite/issues/282)

In /var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml, it has 5 nodes. But 'microk8s.kubectl get nodes -o wide' only shows 4 nodes. I cannot remove the extra node by running `sudo microk8s remove-node node_name` since the node name does not exist. The last node is in 'context deadline exceeded' status.  Do you know how could I remove that node and try to rejoin? Thanks!

**Comment by balchua** [comment]
i know `remove-node` has a `--force` option.

**Comment by rockkoca** [comment]
> i know `remove-node` has a `--force` option.

It only can use the node name to remove a node. At that point, that node name does not exist in the cluster although the node ip address in the cluster.yaml. This might the reason why the api-server keeps restarting. After I gave the new node a different ip address, it joined without any problem.

**Comment by balchua** [comment]
Aaa you're right.  It does a `kubectl get no`.  I don't know if editing the `cluster.yaml` file is possible.

**Comment by rockkoca** [comment]
> Aaa you're right. It does a `kubectl get no`.
> 
> You can actually forcefully remove it from dqlite with this command.
> 
> ```
> snap/microk8s/current/bin/dqlite -s file:///var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml -c /var/snap/microk8s/current/var/kubernetes/backend/cluster.crt -k /var/snap/microk8s/current/var/kubernetes/backend/cluster.key -f json k8s ".remove <node-ip-with-port-19001>"
> ```

It's removed! Thanks!

**Comment by rockkoca** [comment]
Although it's removed, the  'context deadline exceeded' still happens for this specific IP address.

Only 3 nodes show up in cluster.yaml while I have 6 nodes now including this failed one. And, all IDs are 0 in this newly joined node's cluster.yaml file.
The cluster.yaml files in other good nodes have the correct info. But `kubectl get no` does not show the newly joined node since it cannot sync.

`- Address: xxx:19001
  ID: 0
  Role: 0
- Address: xxx:19001
  ID: 0
  Role: 0
- Address: xxx:19001
  ID: 0
  Role: 0`
